src/Microsoft.AspNet.Mvc.Core/Formatters/OutputFormatter.cs

```diff
@@ -268,12 +294,20 @@ protected string GetMediaTypeWithCharset(string mediaType, Encoding encoding)
             }
             else if (mediaTypeEncoding == null)
             {
-                return CreateMediaTypeWithEncoding(mediaType, encoding);
+                if (encoding == Encoding.UTF8)
+                {
+                    return OutputMediaTypeCache[mediaType];
+                }
+                else
+                {
+                    return CreateMediaTypeWithEncoding(mediaType, encoding);
+                }
             }
             else
             {
                 // This can happen if the user has overriden SelectCharacterEncoding
-                return MediaTypeEncoding.ReplaceEncoding(mediaType, encoding);
+                var result = GetFromCacheIfCompatible(mediaType, encoding);
+                return result != null ? result : MediaTypeEncoding.ReplaceEncoding(mediaType, encoding);
```

Comment:
I think your decomposition of these cases is wrong.

Both the `else if` and `else` handle cases where the value comes from the cache. Both the `else if` and `else` handle cases where the string needs to change.